// electrom/lib/renderer/PerfBoard.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PerfBoard = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const monitor_js_1 = require("monitor.js");
const BOARD_WIDTH = 120;
const BOARD_HEIGHT = 48;
const TEXT_COLOR = 'rgb(0, 255, 0)';
const BORDER_COLOR = 'rgb(17, 51, 17)';
const PerfBoard = () => {
    const containerRef = (0, react_1.useRef)();
    (0, react_1.useEffect)(() => {
        if (!containerRef.current)
            return;
        const fpsBoard = new monitor_js_1.FPSBoard({
            container: containerRef.current,
            alpha: 1,
            width: BOARD_WIDTH,
            textColor: TEXT_COLOR,
            boardColor: BORDER_COLOR,
            containerStyles: {
                position: 'absolute',
                left: 0,
                top: 0,
            },
        });
        const memoryStats = new monitor_js_1.MemoryStats({
            containerWidth: BOARD_WIDTH,
        });
        memoryStats.domElement.style.position = 'absolute';
        memoryStats.domElement.style.left = '0px';
        memoryStats.domElement.style.top = `${BOARD_HEIGHT + 2}px`;
        containerRef.current.appendChild(memoryStats.domElement);
        const timer = new monitor_js_1.Timer();
        timer.update(() => {
            fpsBoard.tick();
            memoryStats.update();
        });
        timer.start();
    }, [containerRef.current]);
    return ((0, jsx_runtime_1.jsx)("div", { ref: (ref) => (containerRef.current = ref), style: {
            display: 'flex',
            justifyContent: 'flex-end',
        } }));
};
exports.PerfBoard = PerfBoard;
exports.default = exports.PerfBoard;
